ement-room.el

```diff
@@ -2817,151 +2809,113 @@ MSG is the message STRING was contained in."
-          :then (lambda (data)
-                  (ement-room--insert-preview buf msg related-event data num))
-          :else (lambda (why)
-                  (ement-debug "Error in fetching preview: " why)))))))
-
-;; (defun ement-room--insert-preview (buf msg data num)
-;;   "Insert link preview for DATA into MSG, if it isn't already present.
-;; BUF is the room buffer this link is from. NUM is the index for this link."
-;;   (cond ((alist-get 'errcode data) (message "Ement: You are being rate limited."))
-;;         (data (with-current-buffer buf
-;;                 (let* ((node (ement-room--ewoc-node-before ement-ewoc
-;;                                                            msg
-;;                                                            (lambda (a b)
-;;                                                              (string= (ement-event-id a)
-;;                                                                       (ement-event-id b)))
-;;                                                            :pred #'ement-event-p))
-;;                        (next? (cl-loop with w = (ewoc-next ement-ewoc node)
-;;                                        repeat num
-;;                                        if (or (not w)
-;;                                               (not (listp (ewoc-data w)))
-;;                                               (not (eq (car (ewoc-data w))
-;;                                                        'url-preview))
-;;                                               (not (equal (cadr (ewoc-data w))
-;;                                                           data)))
-;;                                        return t
-;;                                        do (setf w (ewoc-next ement-ewoc w))
-;;                                        finally return nil)))
-;;                   (when next?
-;;                     (with-silent-modifications
-;;                       (ewoc-enter-after ement-ewoc
-;;                                         node
-;;                                         `(url-preview ,data)))))))))
-
-(defun ement-room--insert-preview (buf msg related data num)
+(defun ement-room--get-preview (msg link num)
+  "Get preview of LINK. MSG is the message this link is from. NUM is the index for this link."
+  (let* ((id (ement-event-id msg))
+         (related-event (cl-loop with match = (make-ement-event :id id)
+                                 for e in (ement-room-timeline ement-room)
+                                 if (ement--events-equal-p match e)
+                                 return e))
+         (saved-buffer (current-buffer)))
+    (unless (or ement-room--preview-stop
+                (map-elt (ement-event-local related-event)
+                         'url-preview-n))
+      (ement-api ement-session "preview_url"
+        :endpoint-category "media" :params `((url ,link))
+        :then (lambda (data)
+                (when (buffer-live-p saved-buffer)
```

Comment:
Rather than saving the buffer here, use the room struct itself in the lambda.  That way, if the room's buffer changes before the callback is called, it won't give an error.

Reply:
Yeah, that was what I originally tried, but it seems plz doesn't call the function in the :then key in the same buffer as it was originally called in, instead it calls it in a *plz-request-curl* buffer. I considered filing a bug for plz, but I assumed it was desired behaviour.

I expect most users of plz would also expect that the function should be called in the original buffer plz was run in, so IMHO that should be changed on plz side.

Reply:
I don't think that would work, because there is no guarantee that that buffer would still exist when the callback is called.  If it's necessary to call the callback in a certain buffer, it's the responsibility of the user to use `with-current-buffer` in the callback.